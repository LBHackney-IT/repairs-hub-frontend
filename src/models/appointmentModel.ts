export type Appointment = {
  date: string
  description: string
  start: string
  end: string
  reason: string
  note: string
  assignedStart: string
  assignedEnd: string
  startedAt: string
  bookingLifeCycleStatus: string
}
