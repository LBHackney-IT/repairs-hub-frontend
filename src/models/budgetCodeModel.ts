export type BudgetCode = {
  externalCostCode: string
  corporateSubjectiveCode: string
  descriptionOfWorks: string
}
