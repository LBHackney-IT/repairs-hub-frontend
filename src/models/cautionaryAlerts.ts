export type CautionaryAlert = {
  type: string
  comments: string
  startDate: string
  endDate: string
}

export type CautionaryAlertsResponse = {
  reference: string
  alerts: CautionaryAlert[]
}
