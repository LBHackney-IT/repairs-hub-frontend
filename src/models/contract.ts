export default interface Contract {
  contractReference: string
  terminationDate: string
  effectiveDate: string
  contractorReference: string
  isRaisable: boolean
  contractorName: string
  sorCost: number
  sorCount: number
}
