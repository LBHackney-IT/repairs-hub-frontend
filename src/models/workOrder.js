import { formatISO, isSameDay } from 'date-fns'
import {
  HIGH_PRIORITY_CODES,
  PRIORITY_CODES_REQUIRING_APPOINTMENTS,
} from '@/utils/helpers/priorities'

import {
  CLOSED_STATUS_DESCRIPTIONS,
  CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES,
} from '@/utils/statusCodes'

export class WorkOrder {
  constructor(workOrderData) {
    Object.assign(this, workOrderData)
  }

  isHigherPriority = () => {
    return HIGH_PRIORITY_CODES.includes(this.priorityCode)
  }

  isAppointmentRequired = () => {
    return PRIORITY_CODES_REQUIRING_APPOINTMENTS.includes(this.priorityCode)
  }

  canBeScheduled = () => {
    return (
      this.statusAllowsScheduling() &&
      this.isAppointmentRequired() &&
      !this.isOutOfHoursGas()
    )
  }

  isOutOfHoursGas = () => {
    const GasBreakdownContractorReference = 'H04'
    const OohTradeCode = 'OO'

    var contractorReference = this.contractorReference

    if (contractorReference != GasBreakdownContractorReference) return false // contractor must be "H04"
    var tradeCode = this.tradeCode
    return tradeCode == OohTradeCode
  }

  completionReason = () => {
    return this.status === 'Work Completed' ? 'Completed' : this.status
  }

  appointmentISODatePassed = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    }

    const currentISODate = formatISO(new Date(), { representation: 'date' })

    const appointmentISODate = formatISO(
      new Date(`${this.appointment.date}T${this.appointment.start}`),
      { representation: 'date' }
    )

    return currentISODate >= appointmentISODate
  }

  appointmentIsToday = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    } else {
      return isSameDay(
        new Date(),
        new Date(`${this.appointment.date}T${this.appointment.start}`)
      )
    }
  }

  statusAllowsScheduling = () => {
    return !CLOSED_STATUS_DESCRIPTIONS.includes(this.status)
  }

  targetTimePassed = () => {
    const currentTime = new Date().getTime()
    const targetTime = new Date(`${this.target}`).getTime()

    return currentTime > targetTime
  }

  hasBeenVisited = () =>
    CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES.includes(this.status)
}
