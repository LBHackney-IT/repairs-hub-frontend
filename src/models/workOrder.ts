import { formatISO, isSameDay } from 'date-fns'
import {
  HIGH_PRIORITY_CODES,
  PRIORITY_CODES_REQUIRING_APPOINTMENTS,
} from '@/utils/helpers/priorities'

import {
  CLOSED_STATUS_DESCRIPTIONS,
  CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES,
} from '@/utils/statusCodes'
import { Operative } from './Operative'
import { Appointment } from './appointmentModel'

export class WorkOrder {
  reference: string
  description: string
  externalAppointmentManagementUrl: string
  startTime: string
  appointment: Appointment
  status: string
  dateRaised: string
  tradeDescription: string
  isFollowOn: boolean
  contractorReference: string
  tradeCode: string
  priorityCode
  target
  budgetCode
  closedDated
  operatives: Operative[]
  owner
  priority: string
  raisedBy: string
  callerName: string
  callerNumber: string

  constructor(workOrderData) {
    Object.assign(this, workOrderData)
  }

  isHigherPriority = () => {
    return HIGH_PRIORITY_CODES.includes(this.priorityCode)
  }

  isAppointmentRequired = () => {
    return PRIORITY_CODES_REQUIRING_APPOINTMENTS.includes(this.priorityCode)
  }

  canBeScheduled = () => {
    return (
      this.statusAllowsScheduling() &&
      this.isAppointmentRequired() &&
      !this.isOutOfHoursGas()
    )
  }

  isOutOfHoursGas = () => {
    const gasBreakdownContractorReference = 'H04'
    const oohTradeCode = 'OO'

    const contractorReference = this.contractorReference

    if (contractorReference != gasBreakdownContractorReference) return false // contractor must be "H04"
    const tradeCode = this.tradeCode
    return tradeCode == oohTradeCode
  }

  completionReason = () => {
    return this.status === 'Work Completed' ? 'Completed' : this.status
  }

  appointmentISODatePassed = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    }

    const currentISODate = formatISO(new Date(), { representation: 'date' })

    const appointmentISODate = formatISO(
      new Date(`${this.appointment.date}T${this.appointment.start}`),
      { representation: 'date' }
    )

    return currentISODate >= appointmentISODate
  }

  appointmentIsToday = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    } else {
      return isSameDay(
        new Date(),
        new Date(`${this.appointment.date}T${this.appointment.start}`)
      )
    }
  }

  statusAllowsScheduling = () => {
    return !CLOSED_STATUS_DESCRIPTIONS.includes(this.status)
  }

  targetTimePassed = () => {
    const currentTime = new Date().getTime()
    const targetTime = new Date(`${this.target}`).getTime()

    return currentTime > targetTime
  }

  hasBeenVisited = () =>
    CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES.includes(this.status)
}
