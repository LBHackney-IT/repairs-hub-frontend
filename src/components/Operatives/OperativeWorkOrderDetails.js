import PropTypes from 'prop-types'
import BackButton from '../Layout/BackButton'
import { longMonthWeekday } from '../../utils/date'
import { WorkOrder } from '../../models/workOrder'
import { GridColumn, GridRow } from '../Layout/Grid'
import { useRouter } from 'next/router'
import { useCallback, useState } from 'react'
import { getCautionaryAlertsType } from '../../utils/cautionaryAlerts'

const OperativeWorkOrderDetails = ({
  property,
  workOrder,
  personAlerts,
  locationAlerts,
}) => {
  const router = useRouter()
  const [textOverflow, setTextOverflow] = useState(false)
  const [isExpanded, setIsExpanded] = useState(false)
  const cautionaryAlertsType = getCautionaryAlertsType(
    locationAlerts,
    personAlerts
  )

  const cautContactURL = () => {
    router.push({
      pathname: `/work-orders/cautionary-alerts`,
      query: {
        cautContactCodes: cautionaryAlertsType,
      },
    })
  }

  const onShowMoreClick = () => {
    setIsExpanded(true)
  }

  const onShowLessClick = () => {
    setIsExpanded(false)
  }

  const pRef = useCallback((node) => {
    if (node != null) {
      if (node.scrollHeight > node.clientHeight) {
        setTextOverflow(true)
      }
    }
  })

  return (
    <>
      <div className="operative-work-order">
        <div className="govuk-panel govuk-panel--confirmation lbh-panel appointment-info">
          {workOrder.appointment ? (
            <h3 className="lbh-heading-h3">
              {longMonthWeekday(workOrder.appointment.date)}
              <br />
              {workOrder.appointment.start}-{workOrder.appointment.end}
            </h3>
          ) : (
            <h3 className="lbh-heading-h3">No appointment</h3>
          )}
        </div>
        <BackButton />

        <GridRow>
          <GridColumn width="one-half">
            <div className="lbh-heading-h3">
              WO {workOrder.reference.toString().padStart(8, '0')}
            </div>
          </GridColumn>
          <GridColumn
            width="one-half"
            className="align-grid-column priority-text"
          >
            {workOrder.isHigherPriority() ? (
              <div className="text-dark-red lbh-heading-h3">
                {' '}
                {workOrder.priority.toLowerCase().split(' ').slice(-1)}
              </div>
            ) : (
              <div className="lbh-heading-h3">
                {' '}
                {workOrder.priority.toLowerCase().split(' ').slice(-1)}
              </div>
            )}
          </GridColumn>
        </GridRow>
        <div className="lbh-heading-h5">Description</div>
        <p
          ref={pRef}
          className={`govuk-body govuk-!-margin-bottom-0 ${
            isExpanded ? '' : 'truncate'
          }`}
        >
          {workOrder.description}
        </p>
        {textOverflow ? (
          !isExpanded ? (
            <a
              className="govuk-body lbh-link"
              href="#"
              onClick={onShowMoreClick}
            >
              show more
            </a>
          ) : (
            <a
              className="govuk-body lbh-link"
              href="#"
              onClick={onShowLessClick}
            >
              show less
            </a>
          )
        ) : (
          ''
        )}
        {workOrder.plannerComments && (
          <>
            <div className="lbh-heading-h5">Planner Comment</div>
            <p className="govuk-body">{workOrder.plannerComments}</p>
          </>
        )}

        {cautionaryAlertsType && (
          <GridRow>
            <GridColumn width="one-half">
              <a
                className="lbh-heading-h5 lbh-link"
                href="#"
                id="caut-alerts"
                onClick={cautContactURL}
              >
                Caut. alerts
              </a>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="govuk-warning-text lbh-warning-text">
                <span
                  className="govuk-warning-text__icon person-alert--icon"
                  aria-hidden="true"
                >
                  !
                </span>
                <strong className="govuk-warning-text__text person-alert--text">
                  {cautionaryAlertsType}
                </strong>
              </div>
            </GridColumn>
          </GridRow>
        )}

        <GridRow>
          <GridColumn width="one-half">
            <div className="lbh-body property-name">Address</div>
          </GridColumn>
          <GridColumn width="one-half" className="align-grid-column">
            <div className="lbh-body">
              {property.address.shortAddress} {property.address.postalCode}
            </div>
          </GridColumn>
        </GridRow>

        {workOrder.callerName && (
          <GridRow>
            <GridColumn width="one-half">
              <div className="lbh-body property-name">Contact</div>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="lbh-body">{workOrder.callerName}</div>
            </GridColumn>
          </GridRow>
        )}

        {workOrder.callerNumber && (
          <GridRow>
            <GridColumn width="one-half">
              <div className="lbh-body property-name">Telephone no</div>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="lbh-body">{workOrder.callerNumber}</div>
            </GridColumn>
          </GridRow>
        )}

        {workOrder.appointment && workOrder.appointment.note && (
          <GridRow>
            <GridColumn width="one-half">
              <div className="lbh-body property-name">Comment</div>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="lbh-body">{workOrder.appointment.note}</div>
            </GridColumn>
          </GridRow>
        )}
      </div>
    </>
  )
}

OperativeWorkOrderDetails.propTypes = {
  property: PropTypes.object.isRequired,
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
  personAlerts: PropTypes.array.isRequired,
  locationAlerts: PropTypes.array.isRequired,
}

export default OperativeWorkOrderDetails
