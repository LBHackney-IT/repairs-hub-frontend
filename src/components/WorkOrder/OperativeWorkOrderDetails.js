import PropTypes from 'prop-types'
import WorkOrderHeader from './WorkOrderHeader'
import BackButton from '../Layout/BackButton/BackButton'
import { longMonthWeekday } from '../../utils/date'
import { WorkOrder } from '../../models/workOrder'

const WorkOrderDetails = ({
  property,
  workOrder,
  locationAlerts,
  personAlerts,
  tenure,
  schedulerSessionId,
}) => {
  return (
    <>
      <div class="govuk-panel govuk-panel--confirmation lbh-panel appointment-info">
        <h3 class="lbh-heading-h2">
          {longMonthWeekday(workOrder.appointment.date)}
          <br />
          {workOrder.appointment.start}-{workOrder.appointment.end}
        </h3>
      </div>
      <div className="govuk-!-display-none-print">
        <BackButton />

        <div width="two-thirds">
          <h2 className="lbh-heading-h2 display-inline govuk-!-margin-right-6">
            WO {workOrder.reference.toString().padStart(8, '0')}
          </h2>
        </div>
        <div width="one-third">
          {workOrder.isHigherPriority() ? (
            <h2 className=" text-dark-red lbh-heading-h2 display-inline govuk-!-margin-right-6">
              {workOrder.priority.split(' ').slice(-1)}
            </h2>
          ) : (
            <h2 className="lbh-heading-h2 display-inline govuk-!-margin-right-6">
              {workOrder.priority.split(' ').slice(-1)}
            </h2>
          )}
        </div>

        <p className="lbh-body-m">{workOrder.description}</p>

        <WorkOrderHeader
          propertyReference={property.propertyReference}
          workOrder={workOrder}
          address={property.address}
          locationAlerts={locationAlerts}
          personAlerts={personAlerts}
          subTypeDescription={property.hierarchyType.subTypeDescription}
          tenure={tenure}
          hasLinkToProperty={true}
          canRaiseRepair={property.canRaiseRepair}
          schedulerSessionId={schedulerSessionId}
        />
      </div>
    </>
  )
}

WorkOrderDetails.propTypes = {
  property: PropTypes.object.isRequired,
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
  locationAlerts: PropTypes.array.isRequired,
  personAlerts: PropTypes.array.isRequired,
  tenure: PropTypes.object.isRequired,
}

export default WorkOrderDetails
