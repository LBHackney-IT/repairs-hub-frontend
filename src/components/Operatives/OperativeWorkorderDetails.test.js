import { render } from '@testing-library/react'
import OperativeWorkOrderDetails from './OperativeWorkOrderDetails'
import { WorkOrder } from '@/models/workOrder'

describe('OperativeWorkOrderDetails component', () => {
  let workOrderData = {
    reference: 10000621,
    dateRaised: '2021-06-11T13:49:15.878796Z',
    lastUpdated: null,
    priority: '5 [N] Normal',
    property: '17 Pitcairn House  St Thomass Square',
    owner: 'Herts Heritage Ltd',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    callerName: 'Test Testerson',
    callerNumber: '07856432',
    propertyReference: '00023405',
    tradeCode: 'PL',
    tradeDescription: 'Plumbing - PL',
    status: 'In Progress',
    plannerComment: 'planner comment',
    appointment: {
      date: '2021-09-03',
      description: 'AM Slot',
      start: '08:00',
      end: '13:00',
      reason: null,
      note: 'School run',
    },
  }

  const props = {
    property: {
      propertyReference: '00012345',
      address: {
        shortAddress: '16 Pitcairn House  St Thomass Square',
        postalCode: 'E9 6PT',
        addressLine: '16 Pitcairn House',
        streetSuffix: 'St Thomass Square',
      },
      hierarchyType: {
        levelCode: '7',
        subTypeCode: 'DWE',
        subTypeDescription: 'Dwelling',
      },
      canRaiseRepair: true,
    },
    alerts: {
      personAlert: [
        {
          type: 'DIS',
          comments: 'Property Under Disrepair',
          startDate: '2011-08-16',
          endDate: null,
        },
      ],
      locationAlerts: [
        {
          type: 'CIT',
          comments: '[Temporary] No Lone Interviews',
          startDate: '2011-08-16',
          endDate: null,
        },
      ],
    },
  }

  describe('when on operative-index-page', () => {
    it('should render properly work order', () => {
      const { asFragment } = render(
        <OperativeWorkOrderDetails
          property={props.property}
          workOrder={new WorkOrder(workOrderData)}
          personAlerts={props.alerts.personAlert}
          locationAlerts={props.alerts.locationAlerts}
        />
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
