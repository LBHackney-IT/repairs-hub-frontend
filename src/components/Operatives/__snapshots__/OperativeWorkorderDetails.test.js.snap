// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperativeWorkOrderDetails component when on operative-index-page should render properly work order 1`] = `
<DocumentFragment>
  <div
    class="govuk-panel govuk-panel--confirmation lbh-panel appointment-info"
  >
    <h3
      class="lbh-heading-h3"
    >
      Friday, 3 September
      <br />
      08:00-13:00
    </h3>
  </div>
  <a
    class="govuk-back-link lbh-back-link govuk-!-display-none-print"
    role="button"
  >
    Back
  </a>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        WO 10000621
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="lbh-body"
      >
         Normal
      </div>
    </div>
  </div>
  <div
    class="lbh-heading-h3"
  >
    Description
  </div>
  <p
    class="govuk-body"
  >
    r
  </p>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        Caut. contact
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="govuk-warning-text lbh-warning-text"
      >
        <span
          aria-hidden="true"
          class="govuk-warning-text__icon person-alert--icon"
        >
          !
        </span>
        <strong
          class="govuk-warning-text__text person-alert--text"
        >
          DIS
        </strong>
      </div>
    </div>
  </div>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        Address
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="lbh-body"
      >
        16 Pitcairn House  St Thomass Square E9 6PT
      </div>
    </div>
  </div>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        Contact
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="lbh-body"
      >
        Test Testerson
      </div>
    </div>
  </div>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        Telephone no
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="lbh-body"
      >
        07856432
      </div>
    </div>
  </div>
  <div
    class="govuk-grid-row undefined"
  >
    <div
      class="govuk-grid-column-one-half"
    >
      <div
        class="lbh-heading-h3"
      >
        Comment
      </div>
    </div>
    <div
      class="govuk-grid-column-one-half align-grid-column"
    >
      <div
        class="lbh-body"
      >
        School run
      </div>
    </div>
  </div>
</DocumentFragment>
`;
